src/main/java/net/neoforged/neoforge/transfer/resources/ItemResource.java

```diff
@@ -0,0 +1,310 @@
+     * of the fact it is already an instance with a data component map.
+     */
+    private final ItemStack innerStack;
+
+    private ItemResource(ItemStack stack) {
+        this.innerStack = stack;
+    }
+
+    @Override
+    public Item value() {
+        return innerStack.getItem();
+    }
+
+    /**
+     * @return The {@link Item} of the resource from the inner {@link ItemStack}
+     */
+    public Item getItem() {
+        return value();
+    }
+
+    @Override
+    public Holder<Item> getHolder() {
+        return innerStack.getItemHolder();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return innerStack.isEmpty();
+    }
+
+    /**
+     * {@return true if the stack components and instance matches the inner stack's components and instance} Uses the {@link ItemStack#isSameItemSameComponents(ItemStack, ItemStack)} method for comparison.
+     *
+     * @param stack the item stack to check
+     */
+    public boolean is(ItemStack stack) {
+        return ItemStack.isSameItemSameComponents(stack, innerStack);
+    }
+
+    /**
+     * {@return true if the item instance matches the backing instance value}
+     *
+     * @param item the item to check
+     */
+    public boolean is(ItemLike item) {
+        return is(item.asItem());
+    }
+
+    /**
+     * Tests an {@link ItemStack} predicate with the inner stack.
+     *
+     * @param predicate Predicate to perform the test with
+     * @return {@code true} if the test passed
+     */
+    public boolean test(Predicate<ItemStack> predicate) {
+        return predicate.test(innerStack);
+    }
+
+    @Override
+    public boolean isComponentsPatchEmpty() {
+        return innerStack.isComponentsPatchEmpty();
+    }
+
+    @Override
+    public ItemResource withPatch(DataComponentPatch patch) {
+        if (isEmpty() || patch.isEmpty() || innerStack.getComponentsPatch().equals(patch))
+            return this;
+
+        ItemStack stack = innerStack.copy();
+        stack.applyComponents(patch);
+
+        return ItemResource.of(stack);
+    }
+
+    @Override
+    public <D> ItemResource with(DataComponentType<D> type, D data) {
+        if (isEmpty()) return ItemResource.EMPTY;
+        ItemStack stack = innerStack.copy();
+        stack.set(type, data);
+        if (ItemStack.isSameItemSameComponents(innerStack, stack)) return this;
+
+        return ItemResource.of(stack);
+    }
+
+    //This is overridden to return ItemResource to allow method chaining
+    @Override
+    public <D> ItemResource with(Supplier<? extends DataComponentType<D>> type, D data) {
+        return with(type.get(), data);
+    }
+
+    @Override
+    public ItemResource without(DataComponentType<?> type) {
+        if (isEmpty()) return ItemResource.EMPTY;
+        ItemStack stack = innerStack.copy();
+        stack.remove(type);
+        if (ItemStack.isSameItemSameComponents(innerStack, stack)) return this;
```

Comment:
Just to streamline things a bit, even though it relies on implementation details on how data component maps (and patches) treat removals, instead of checking *all* the components again, before even copying the stack. I believe we could just make it short circuit instead
```java
if (innerStack.get(type) == null) return this;
```